### PR TITLE
server: make SQL connection keep alive easily configurable

### LINKED ISSUE #115422: sql: detect broken TCP client connections sooner
If a TCP connection is abandoned without being explicitly closed by the remote peer via a TCP RST packet, we rely on TCP keepalives to probe the connection and close it after some time of inactivity. This can happen e.g. if the remote host crashes (including the OS), if the network link stops sending traffic, or if some network device drops the connection without closing it.

If this is not detected promptly, it can cause a transaction to suddenly halt while open. This is dangerous if the transaction has taken out locks (i.e. made any writes), since it will block other transactions from accessing the locked rows until it's closed and rolled back, causing contention pileups that last for several minutes.

We need to detect such broken TCP connections sooner and close them, such that the transactions are rolled back and the locks released.

By default, the Linux kernel will probe the connection only after 2 hours of inactivity, and then send 9 probes every 75 seconds before closing the connection if no response is received:

```
# NB: roachprod clusters use different defaults
$ sysctl -a | grep tcp_keepalive
net.ipv4.tcp_keepalive_intvl = 75
net.ipv4.tcp_keepalive_probes = 9
net.ipv4.tcp_keepalive_time = 7200
```

However, we set an explicit 1 minute keepalive time and interval on the socket:

https://github.com/cockroachdb/cockroach/blob/c0d5d08756a9ac2b7a92619b6fbb209f6e1fc44c/pkg/server/tcp_keepalive_manager.go#L56-L68

https://github.com/cockroachdb/cockroach/blob/c0d5d08756a9ac2b7a92619b6fbb209f6e1fc44c/pkg/server/tcp_keepalive_manager.go#L34

But this only sets `TCP_KEEPIDLE` and `TCP_KEEPINTVL` to 1 minute each (i.e. the time before the first probe and the interval between probes), while the number of probes `TCP_KEEPCNT` is not modified and remains at its default 9. So it takes as long as 10 minutes before we detect the connection failure and close it.

We should:

1. Reconsider the value of `TCP_KEEPIDLE` and `TCP_KEEPINTVL` to be more aggressive.
2. Set `TCP_KEEPCNT` to a lower value, and make it configurable.
3. Consider making `COCKROACH_SQL_TCP_KEEP_ALIVE` a cluster setting, if we can easily do so.

Something like `TCP_KEEPIDLE=20s`, `TCP_KEEPINTVL=20s`, `TCP_KEEPCNT=3` might make sense. This will close the connection after 1 minute, having sent 3 probes.

## Repro

Here's a repro showing the transaction holding locks for 10 minutes after the connection is broken (by stopping traffic with iptables):

```
roachprod create grinaker-tcp -n 3
roachprod run grinaker-tcp 'sudo sysctl net.ipv4.tcp_keepalive_time=7200 net.ipv4.tcp_keepalive_probes=9 net.ipv4.tcp_keepalive_intvl=75'
roachprod stage grinaker-tcp release v22.2.13
roachprod start grinaker-tcp
roachprod sql grinaker-tcp:1

create table test (id int primary key, value string not null);
insert into test values (1, 'a'), (2, 'b'), (3, 'c');
```

Now, create a separate SQL session across the network from n1 to n2, begin a transaction, and update row 2 to take out a lock:

```
$ roachprod ip grinaker-tcp:2
10.142.1.183
$ roachprod ssh grinaker-tcp:1
$ cockroach sql --url postgres://root@10.142.1.183:26257?sslmode=disable
> begin;
OPEN> update test set value = 'foo' where id = 2;
```

Do not close this session. Open a new session, trigger lock contention by reading the table, and inspect the locks and transaction:

```
$ roachprod sql grinaker-tcp:1
> select * from test;
^C
attempting to cancel query...
ERROR: query execution canceled
SQLSTATE: 57014

> \x
> select * from crdb_internal.cluster_locks;
-[ RECORD 1 ]
range_id        | 53
table_id        | 104
database_name   | defaultdb
schema_name     | public
table_name      | test
index_name      | 
lock_key        | \xf0898a88
lock_key_pretty | /Table/104/1/2/0
txn_id          | 1f9d294e-19a8-402d-97f7-6afc7d6ffab0
ts              | 2023-12-01 14:02:04.940638
lock_strength   | Exclusive
durability      | Replicated
granted         | t
contended       | f
duration        | 00:00:52.908026

> select * from crdb_internal.cluster_transactions where id = '1f9d294e-19a8-402d-97f7-6afc7d6ffab0';
-[ RECORD 1 ]
id                     | 1f9d294e-19a8-402d-97f7-6afc7d6ffab0
node_id                | 2
session_id             | 179cb9eda2f914080000000000000002
start                  | 2023-12-01 14:00:19.97722
txn_string             | "sql txn" meta={id=1f9d294e key=/Table/104/1/2/0 pri=0.00467454 epo=0 ts=1701439324.940637758,2 min=1701439219.977209358,0 seq=1} lock=true stat=PENDING rts=1701439219.977209358,0 wto=false gul=1701439220.477209358,0
application_name       | $ cockroach sql
num_stmts              | 3
num_retries            | 0
num_auto_retries       | 0
last_auto_retry_reason | 

> select * from crdb_internal.cluster_sessions where kv_txn = '1f9d294e-19a8-402d-97f7-6afc7d6ffab0';
-[ RECORD 1 ]
node_id            | 2
session_id         | 179cb9eda2f914080000000000000002
user_name          | root
client_address     | 10.142.1.187:50232
application_name   | $ cockroach sql
active_queries     | 
last_active_query  | SHOW database
num_txns_executed  | 4
session_start      | 2023-12-01 13:59:49.531789
active_query_start | NULL
kv_txn             | 1f9d294e-19a8-402d-97f7-6afc7d6ffab0
alloc_bytes        | 40960
max_alloc_bytes    | 40960
status             | IDLE
session_end        | NULL
```

Note the client address 10.142.1.187:50232, which is from n1. Let's break the connection by dropping all packets sent across it:

```
roachprod ssh grinaker-tcp:1
sudo iptables -A OUTPUT -p tcp --sport 50232 -j DROP
sudo iptables -A INPUT -p tcp --dport 50232 -j DROP
```

Observe the transaction still remaining open and holding locks for the next 10 minutes:

```
> select *, now() - start as duration from crdb_internal.cluster_transactions where id = '1f9d294e-19a8-402d-97f7-6afc7d6ffab0';
-[ RECORD 1 ]
id                     | 1f9d294e-19a8-402d-97f7-6afc7d6ffab0
node_id                | 2
session_id             | 179cb9eda2f914080000000000000002
start                  | 2023-12-01 14:00:19.97722
txn_string             | "sql txn" meta={id=1f9d294e key=/Table/104/1/2/0 pri=0.00467454 epo=0 ts=1701439324.940637758,2 min=1701439219.977209358,0 seq=1} lock=true stat=PENDING rts=1701439219.977209358,0 wto=false gul=1701439220.477209358,0
application_name       | $ cockroach sql
num_stmts              | 3
num_retries            | 0
num_auto_retries       | 0
last_auto_retry_reason | 
duration               | 00:10:44.790427

> select * from crdb_internal.cluster_locks;                                                                                    
-[ RECORD 1 ]
range_id        | 53
table_id        | 104
database_name   | defaultdb
schema_name     | public
table_name      | test
index_name      | 
lock_key        | \xf0898a88
lock_key_pretty | /Table/104/1/2/0
txn_id          | 1f9d294e-19a8-402d-97f7-6afc7d6ffab0
ts              | 2023-12-01 14:02:04.940638
lock_strength   | Exclusive
durability      | Replicated
granted         | t
contended       | f
duration        | 00:08:18.424636
```

Jira issue: CRDB-34025

Epic CRDB-35306

### PR DESCRIPTION
**Pull request by fqazi**
Previously, the only way to control the keep-alive time for CRDB was through the environment variable COCKROACH_SQL_TCP_KEEP_ALIVE, which set the keep-alive idle and probe times only. This was difficult to use, and the default meant that CRDB kept connections alive for 10 minutes of idle time. To help make this more configurable, we will introduce two cluster settings:  server.sql_tcp_keep_alive_probe_interval and server.sql_tcp_keep_alive_timeout. These settings will control the keep-alive probe / idle intervals, and the probe count (derived from the timeout) before the connection will be dropped.

Fixes: #115422

### CODE REVIEW

**Review by erikgrinaker — COMMENTED** [review]
This change may be too large to backport. Are we planning a more targeted fix for backports?

> Reviewed 5 of 5 files at r1, 8 of 8 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115833#-:-NlN8dWE6wIP5KLLSRty:b-ocqjre)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @abarganier and @fqazi)

___
*[`-- commits` line 27 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/115833#-NlNAIvw80eJJs7V5gTM):*
This also changes the default values, and removes the env var -- we should call that out as well.
___
*[`pkg/server/tcp_keepalive_manager.go` line 28 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/115833#-NlNAeky9oAVPH9Fw9ws:-NlNAekzA_-OL8pLZJ1N:b-5q8n9j) ([raw file](https://github.com/cockroachdb/cockroach/blob/c753ad9e7fad848af5605c1b4f50e2fc19827665/pkg/server/tcp_keepalive_manager.go#L28)):*
> ```Go
> var KeepAliveTimeout = settings.RegisterDurationSetting(
> 	settings.ApplicationLevel,
> 	"server.sql_tcp_keep_alive.timeout",
> ```

Did you consider simply exposing the parameters directly, [like Postgres does](https://www.postgresql.org/docs/current/runtime-config-connection.html#RUNTIME-CONFIG-TCP-SETTINGS)?
___
*[`pkg/server/tcp_keepalive_manager.go` line 29 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/115833#-NlN9bMwBDOLL1DohEww:-NlN9bMwBDOLL1DohEwx:b-t35lnl) ([raw file](https://github.com/cockroachdb/cockroach/blob/c753ad9e7fad848af5605c1b4f50e2fc19827665/pkg/server/tcp_keepalive_manager.go#L29)):*
> ```Go
> 	settings.ApplicationLevel,
> 	"server.sql_tcp_keep_alive.timeout",
> 	"maximum time for which an idle TCP connection to the server will be kept alive",
> ```

Consider "unresponsive" instead of "idle". It's important to distinguish this from the case of a connected live client that's simply not doing anything.
___
*[`pkg/server/tcp_keepalive_manager.go` line 30 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/115833#-NlN9lup9-x4lVWhj3jL:-NlN9lup9-x4lVWhj3jM:bpw03xm) ([raw file](https://github.com/cockroachdb/cockroach/blob/c753ad9e7fad848af5605c1b4f50e2fc19827665/pkg/server/tcp_keepalive_manager.go#L30)):*
> ```Go
> 	"server.sql_tcp_keep_alive.timeout",
> 	"maximum time for which an idle TCP connection to the server will be kept alive",
> 	time.Minute*2,
> ```

2 minutes seems very high here. I think the only benefit of a high timeout is to tolerate high network latency and transient disconnects, while the downsides can be quite severe (workload stall). Is it important to handle a 2 minute network latency or disconnect?
___
*[`pkg/server/tcp_keepalive_manager.go` line 79 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/115833#-NlNB2dr9w0BxxpUCCSu:-NlNB2dr9w0BxxpUCCSv:b-4abgtu) ([raw file](https://github.com/cockroachdb/cockroach/blob/c753ad9e7fad848af5605c1b4f50e2fc19827665/pkg/server/tcp_keepalive_manager.go#L79)):*
> ```Go
> 	// probe count.
> 	maximumConnectionTime := KeepAliveTimeout.Get(&k.settings.SV)
> 	probeFrequency := KeepAliveProbeFrequency.Get(&k.settings.SV)
> ```

We don't guard against either of these being set to 0 (`KeepAliveTimeout = 0` will result in division by 0 below).

We should either take 0 to mean no probes are sent (like in Postgres), or disallow it.
___
*[`pkg/server/tcp_keepalive_manager_test.go` line 31 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/115833#-NlNBLjADW0o2L5ClAFZ:-NlNBLjADW0o2L5ClAF_:be9uh5p) ([raw file](https://github.com/cockroachdb/cockroach/blob/c753ad9e7fad848af5605c1b4f50e2fc19827665/pkg/server/tcp_keepalive_manager_test.go#L31)):*
> ```Go
> )
> 
> func TestKeepAliveManager(t *testing.T) {
> ```

We should have an end-to-end test as well, which tests that an unresponsive SQL client connection is in fact disconnected and its session torn down within the expected time.

**Review by fqazi — COMMENTED** [review]
I'm tempted to say we just narrow down the bits for linux and make them an environment variable enabled by default. What do you think?

@rafiss @erikgrinaker RFAL

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115833#-:-NnofuJJDn4wuE9X23DG:bm5heqe)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @erikgrinaker, and @renatolabs)

___
*[`pkg/server/tcp_keepalive_manager.go` line 28 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/115833#-NlNAeky9oAVPH9Fw9ws:-NnjKpVxBQ6Pv3vTxG7M:br3fmx3) ([raw file](https://github.com/cockroachdb/cockroach/blob/c753ad9e7fad848af5605c1b4f50e2fc19827665/pkg/server/tcp_keepalive_manager.go#L28)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Did you consider simply exposing the parameters directly, [like Postgres does](https://www.postgresql.org/docs/current/runtime-config-connection.html#RUNTIME-CONFIG-TCP-SETTINGS)?
</blockquote></details>

Yeah, lets keep this simple, and I named it as:
server.sql_tcp_keep_alive.count
___
*[`pkg/server/tcp_keepalive_manager.go` line 29 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/115833#-NlN9bMwBDOLL1DohEww:-NnjKTKY5Z7ZoGVaCX1w:b-459v3s) ([raw file](https://github.com/cockroachdb/cockroach/blob/c753ad9e7fad848af5605c1b4f50e2fc19827665/pkg/server/tcp_keepalive_manager.go#L29)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Consider "unresponsive" instead of "idle". It's important to distinguish this from the case of a connected live client that's simply not doing anything.
</blockquote></details>

Good point changed this to:

maximum number of probes that will be sent out before a connection dropped because its unresponsive (Linux and Darwin only)
___
*[`pkg/server/tcp_keepalive_manager.go` line 30 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/115833#-NlN9lup9-x4lVWhj3jL:-Nnofn0EBWKdzyHoTZ-6:b-flwfi0) ([raw file](https://github.com/cockroachdb/cockroach/blob/c753ad9e7fad848af5605c1b4f50e2fc19827665/pkg/server/tcp_keepalive_manager.go#L30)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

2 minutes seems very high here. I think the only benefit of a high timeout is to tolerate high network latency and transient disconnects, while the downsides can be quite severe (workload stall). Is it important to handle a 2 minute network latency or disconnect?
</blockquote></details>

I reduced this to a minute, let me know what your gut feel is for this setting.

**Review by erikgrinaker — APPROVED** [review]
> Reviewed 5 of 5 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115833)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @fqazi, and @renatolabs)

___
*[`pkg/cmd/roachtest/tests/network.go` line 333 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/115833#-Nnt3Pzo146ij5pXshVs:-Nnt3Pzo146ij5pXshVt:b-jxw638) ([raw file](https://github.com/cockroachdb/cockroach/blob/e233d7f5cf75362222a767600afc58f0bb287647/pkg/cmd/roachtest/tests/network.go#L333)):*
> ```Go
> 			return err
> 		}
> 		// We expect 2 since the query will itself
> ```

nit: comment seems stale (count != 1)
___
*[`pkg/cmd/roachtest/tests/network.go` line 349 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/115833#-Nnt3wurElKUGsHdPsCI:-Nnt3wurElKUGsHdPsCJ:b-u4g0wm) ([raw file](https://github.com/cockroachdb/cockroach/blob/e233d7f5cf75362222a767600afc58f0bb287647/pkg/cmd/roachtest/tests/network.go#L349)):*
> ```Go
> 
> # Drop any client traffic to CRDB.
> sudo iptables -A INPUT -p tcp --dport 26257 -j DROP;
> ```

This should be --sport 26257, not --dport. The traffic is coming from the CRDB process on the remote node (so its source port is 26257), to a random client port (dport) on this node.

As is, the client will receive packets from the server, but can't send packets to it. That still works, but we may as well sever the connection completely.
___
*[`pkg/cmd/roachtest/tests/network.go` line 352 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/115833#-Nnt3Yhu1Tf8lqrqH0CY:-Nnt3Yhu1Tf8lqrqH0CZ:bhgotoe) ([raw file](https://github.com/cockroachdb/cockroach/blob/e233d7f5cf75362222a767600afc58f0bb287647/pkg/cmd/roachtest/tests/network.go#L352)):*
> ```Go
> sudo iptables -A OUTPUT -p tcp --dport 26257 -j DROP;
> 
> sudo iptables-save
> ```

nit: does this have any purpose? This just saves the ruleset to disk in case the host restarts, the rules are already applied.
___
*[`pkg/cmd/roachtest/tests/network.go` line 363 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/115833#-Nnt4UlA0LwMjnm_72A_:-Nnt4UlA0LwMjnm_72Aa:bfqkhkk) ([raw file](https://github.com/cockroachdb/cockroach/blob/e233d7f5cf75362222a767600afc58f0bb287647/pkg/cmd/roachtest/tests/network.go#L363)):*
> ```Go
> 		const restoreNet = `
> set -e;
> sudo iptables -D INPUT -p tcp --dport 26257 -j DROP;
> ```

nit: can also just do `iptables -F INPUT` to remove all rules.
___
*[`pkg/server/tcp_keepalive_manager.go` line 30 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/115833#-NlN9lup9-x4lVWhj3jL:-Nnt51rF-HWaVX5GM-cO:buf1gpx) ([raw file](https://github.com/cockroachdb/cockroach/blob/c753ad9e7fad848af5605c1b4f50e2fc19827665/pkg/server/tcp_keepalive_manager.go#L30)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

I reduced this to a minute, let me know what your gut feel is for this setting.
</blockquote></details>

I think that's still towards the higher end. For comparison, CRDB RPC connections use a keepalive timeout of 10 seconds, and that's only because gRPC doesn't allow lower values for various reasons.

https://github.com/cockroachdb/cockroach/blob/2e6eb2ccd6337594355b3af9acb0bfb0b6f50225/pkg/rpc/keepalive.go#L55-L59

This basically says "how long do we want to allow a client to disconnect from the network and then reconnect without closing its session"? Since we'll be holding onto the transaction's locks in the meanwhile, I'm inclined to make this more aggressive -- it seems better for a client to retry instead if this should happen. Maybe 20-30 seconds? That still feels pretty lenient -- I don't think it's unreasonable to disconnect a client who falls off the network for that long. Then again, it's probably going to break something for someone somewhere, and 1 minute would be the conservative option -- it's still way better than 10 minutes.

I guess there's also a question of how chatty we want to be on idle client connections.

I'll leave this to SQL Foundations' discretion, since you probably know more about client environments and such -- personally I'd lean towards something like 30 seconds, but 1 minute also seems reasonable. Hard to say what the optimal value is.
___
*[`pkg/server/tcp_keepalive_manager.go` line 29 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/115833#-Nnt4wt15RaDXncWAmJb:-Nnt4wt15RaDXncWAmJc:b63bjdu) ([raw file](https://github.com/cockroachdb/cockroach/blob/e233d7f5cf75362222a767600afc58f0bb287647/pkg/server/tcp_keepalive_manager.go#L29)):*
> ```Go
> 	settings.ApplicationLevel,
> 	"server.sql_tcp_keep_alive.count",
> 	"maximum number of probes that will be sent out before a connection dropped because "+
> ```

nit: "a connection is dropped" and "it's unresponsive"

**Review by fqazi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115833)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @erikgrinaker, and @renatolabs)

___
*[`pkg/cmd/roachtest/tests/network.go` line 352 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/115833#-Nnt3Yhu1Tf8lqrqH0CY:-NnzUGnD0x6bQrtIEOAD:bchqavt) ([raw file](https://github.com/cockroachdb/cockroach/blob/e233d7f5cf75362222a767600afc58f0bb287647/pkg/cmd/roachtest/tests/network.go#L352)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

nit: does this have any purpose? This just saves the ruleset to disk in case the host restarts, the rules are already applied.
</blockquote></details>

Done.

Good point, cleaned this up.
___
*[`pkg/server/tcp_keepalive_manager.go` line 30 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/115833#-NlN9lup9-x4lVWhj3jL:-NnzUKuPDts3LRFi5gY2:bsoy1z0) ([raw file](https://github.com/cockroachdb/cockroach/blob/c753ad9e7fad848af5605c1b4f50e2fc19827665/pkg/server/tcp_keepalive_manager.go#L30)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

I think that's still towards the higher end. For comparison, CRDB RPC connections use a keepalive timeout of 10 seconds, and that's only because gRPC doesn't allow lower values for various reasons.

https://github.com/cockroachdb/cockroach/blob/2e6eb2ccd6337594355b3af9acb0bfb0b6f50225/pkg/rpc/keepalive.go#L55-L59

This basically says "how long do we want to allow a client to disconnect from the network and then reconnect without closing its session"? Since we'll be holding onto the transaction's locks in the meanwhile, I'm inclined to make this more aggressive -- it seems better for a client to retry instead if this should happen. Maybe 20-30 seconds? That still feels pretty lenient -- I don't think it's unreasonable to disconnect a client who falls off the network for that long. Then again, it's probably going to break something for someone somewhere, and 1 minute would be the conservative option -- it's still way better than 10 minutes.

I guess there's also a question of how chatty we want to be on idle client connections.

I'll leave this to SQL Foundations' discretion, since you probably know more about client environments and such -- personally I'd lean towards something like 30 seconds, but 1 minute also seems reasonable. Hard to say what the optimal value is.
</blockquote></details>

Done.

Yeah, I dropped this down to ~30 seconds with a probe every 10 seconds. Also, I'm probably being too cautious, since I wouldn't expect that heavy packet loss to a database server.

**Comment by fqazi** [comment]
@erikgrinaker TFTR!

bors r+

**Comment by renatolabs** [comment]
This PR has drifted and the roachtest changes here no longer compile on master because of #117320.

TLDR: The list of nodes to run a command in should be passed to`RunE`  using `option.WithNodes(nodes)`.

bors r-
